app/src/main/java/com/duckduckgo/app/fire/CookieRemover.kt

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2020 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.fire
+
+import android.database.DatabaseErrorHandler
+import android.database.sqlite.SQLiteDatabase
+import android.webkit.CookieManager
+import com.duckduckgo.app.global.DispatcherProvider
+import com.duckduckgo.app.global.exception.UncaughtExceptionRepository
+import com.duckduckgo.app.global.exception.UncaughtExceptionSource
+import com.duckduckgo.app.statistics.pixels.Pixel
+import kotlinx.coroutines.withContext
+import timber.log.Timber
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+interface CookieRemover {
+    suspend fun removeCookies(): Boolean
+}
+
+class CookieManagerRemover(private val cookieManager: CookieManager) : CookieRemover {
+    override suspend fun removeCookies(): Boolean {
+        suspendCoroutine<Unit> { continuation ->
+            cookieManager.removeAllCookies {
+                Timber.v("All cookies removed; restoring DDG cookies")
+                continuation.resume(Unit)
+            }
+        }
+        return true
+    }
+}
+
+class SQLCookieRemover(
+    private val webViewDatabaseLocator: DatabaseLocator,
+    private val getCookieHostsToPreserve: GetCookieHostsToPreserve,
+    private val pixel: Pixel,
+    private val uncaughtExceptionRepository: UncaughtExceptionRepository,
+    private val dispatcherProvider: DispatcherProvider
+) : CookieRemover {
+
+    override suspend fun removeCookies(): Boolean {
+        return withContext(dispatcherProvider.io()) {
+            val databasePath: String = webViewDatabaseLocator.getDatabasePath()
+            if (databasePath.isNotEmpty()) {
+                val excludedHosts = getCookieHostsToPreserve()
+                return@withContext removeCookies(databasePath, excludedHosts)
+            } else {
+                pixel.fire(Pixel.PixelName.COOKIE_DATABASE_NOT_FOUND)
+            }
+            return@withContext false
+        }
+    }
+
+    private suspend fun openReadableDatabase(databasePath: String): SQLiteDatabase? {
+        return try {
+            SQLiteDatabase.openDatabase(
+                databasePath,
+                null,
+                SQLiteDatabase.OPEN_READWRITE,
+                DatabaseErrorHandler { Timber.e("COOKIE: onCorruption") })
+        } catch (exception: Exception) {
+            pixel.fire(Pixel.PixelName.COOKIE_DATABASE_OPEN_ERROR)
```

Comment:
If the app is about to crash, there's a good chance this pixel will not be sent

Reply:
In this case, it's probably fine to use the uncaught exception handler here and not sending your own pixel. one nuance: you are using the same `UncaughtExceptionSource.COOKIE_DATABASE` for all database problems, so maybe you'd need to add different sources for each potential database problem (like you currently have the different cases covered by different pixels)